rfcs/0116-artifact-service.md

```diff
@@ -0,0 +1,196 @@
+# RFC 116 Artifact Service
+* Comments: [#116](https://api.github.com/repos/taskcluster/taskcluster-rfcs/issues/116)
+* Proposed by: @jhford
+
+# Summary
+
+Taskcluster tasks often product Artifacts during the course of being run.
+Examples include browser archives, json metadata and log files.  We currently
+store Artifacts in a single region of S3 and split management of Artifacts
+between the Queue and Cloud Mirror.  This project is to take Artifact handling
+out of the Queue and into its own service that takes over the responsibilities
+for cross-region and cross-cloud replication.
+
+The Artifact service should not be directly exposed to end users as a Rest API.
+It will instead be a generalised API that the Queue uses to facilitate artifact
+creation from tasks.  In other words, the Queue gates operations to the
+Artifact service.  The service will be exposed to endusers by means of a `Location`
+header value on a `300` series redirect.
+
+The Queue will now store only the name of Artifacts and would treat the name of
+an Artifact as an opaque identifier and defer all management of the actual
+bytes to this system.  The Queue would also stop doing IP to Object store
+mappings.
+
+This RFC is to decide on what the Artifacts service is and its interface, not
+implementation details
+
+## Motivation
+
+* We want the Queue to be easier to understand, develop and maintain
+* We want Artifacts to be easier to understand, develop and maintain
+* We want better Artifact handling without increasing complexity of the Queue
+  codebase
+* We want to have more clear responsibilities for our services
+* We want to consolidate the handling of Artifacts into a single service
+* We want to build generalised software when possible as participants of a free
+  software community
+
```

Comment:
Could we add something about how someone who runs their own data center could point the queue at something else that implements the same API as this blob/object storage  service thing..

Reply:
A large portion of CoT's complexity is dedicated to making sure artifacts at rest haven't been altered. If we could save sha metadata with the artifacts which could a) be compared with the uploaded sha, and b) be compared with the downloaded contents, that would go a long way to simplifying our security checks.

Are there any plans to add such information to the artifact service?

Reply:
@escapewindow, the new artifact kind `storageType: 'blob'`, see [createArtifact](https://docs.taskcluster.net/reference/platform/taskcluster-queue/references/api#createArtifact), actually has `contentSha256`. CoT could definitely use this in some way.

It's not immediately obvious to me how we'll do it.. as we'll still need to know which artifacts were covered by the CoT certificate. Since artifacts like task log can't be under the CoT. Or maybe in future we put that under CoT and accept that certain errors will be more silent, that might be okay, since errors computing CoT is an error in the worker... or network error between worker and queue.

Reply:
yep, the service will always do as strong checks as possible to ensure that we allow only uploading of a known object (e.g. same Sha256 value), and that downloads will allow information to ensure that the object is what it should've been when uploaded.  The download verification will need to happen in the client, as there's nothing in the HTTP spec stronger than `Content-MD5` for verification.

I plan to adapt taskcluster-lib-artifact to support this automatic verification when downloading from this service.